CRM/Campaign/BAO/Campaign.php

```diff
@@ -341,6 +343,24 @@ public static function accessCampaign() {
     return $allow;
   }
 
+  /**
+   * @return bool
+   */
+  public static function viewCampaign() {
+    static $allowView = NULL;
+
+    if (!isset($allowView)) {
+      $allowView = FALSE;
+      if (self::accessCampaign() ||
+        CRM_Core_Permission::check('view campaign')
+      ) {
+        $allowView = TRUE;
+      }
+    }
+
+    return $allowView;
+  }
```

Comment:
@colemanw I wouldn't say recently but this (overly cached I think) function https://github.com/civicrm/civicrm-core/blob/fc3ae9de6a9681ee7a72cb06d8d88a38e1f38356/CRM/Core/Permission.php#L924 handles it & I think you could just add to https://github.com/civicrm/civicrm-core/blob/fc3ae9de6a9681ee7a72cb06d8d88a38e1f38356/CRM/Core/Permission.php#L909 although the naming / comments suggest that is a narrow function.

It looks like it was all written with the expectation that a hook might be added to it at some point (which would explain the excessive caching)

Reply:
@eileenmcnaughton instead of adding a new hook, what about allowing an extra key in the arrays returned by `hook_civicrm_permissions`. So for example if we want to sensibly have `delete` imply `edit` we could do:
```diff
--- a/ext/civigrant/civigrant.php
+++ b/ext/civigrant/civigrant.php
@@ -64,6 +64,7 @@ function civigrant_civicrm_permission(&$permissions) {
   $permissions['delete in CiviGrant'] = [
     E::ts('CiviGrant:') . ' ' . E::ts('delete in CiviGrant'),
     E::ts('Delete grants'),
+    'implies' => ['edit grants'],
   ];
 }

```

And while we're at it, those arrays would be more readable if they were associative so we also ought to support this format:
```diff
diff --git a/ext/civigrant/civigrant.php b/ext/civigrant/civigrant.php
index 1f491b57d9..6a87bf29a7 100644
--- a/ext/civigrant/civigrant.php
+++ b/ext/civigrant/civigrant.php
@@ -62,8 +62,9 @@ function civigrant_civicrm_permission(&$permissions) {
     E::ts('Create and update grants'),
   ];
   $permissions['delete in CiviGrant'] = [
-    E::ts('CiviGrant:') . ' ' . E::ts('delete in CiviGrant'),
-    E::ts('Delete grants'),
+    'title' => E::ts('CiviGrant:') . ' ' . E::ts('delete in CiviGrant'),
+    'description' => E::ts('Delete grants'),
+    'implies' => ['edit grants'],
   ];
 }
 

```

Reply:
@colemanw looks like a good plan!

Reply:
This does just that! https://github.com/civicrm/civicrm-core/pull/29174